corehq/apps/sso/forms.py

```diff
@@ -0,0 +1,550 @@
+    if slug != slugified_test:
+        raise forms.ValidationError(
+            _("You did not enter a valid slug. Did you mean '{}'?").format(
+                slugified_test)
+        )
+
+
+def _check_is_editable_requirements(identity_provider):
+    if not identity_provider.email_domains:
+        raise forms.ValidationError(
+            _("Please make sure you specify at "
+              "least one Authenticated Email Domain.")
+        )
+    if not identity_provider.sso_exempt_users:
+        raise forms.ValidationError(
+            _("Please make sure you have specified at least one "
+              "enterprise admin that is exempt from SSO.")
+        )
+
+
+def _check_required_when_active(is_active, value):
+    if is_active and not value:
+        raise forms.ValidationError(
+            _("This is required when Single Sign On is active.")
+        )
+
+
+class CreateIdentityProviderForm(forms.Form):
+    """This form initializes the essential fields of an IdentityProvider
+    """
+    owner = forms.IntegerField(
+        label=ugettext_lazy("Billing Account Owner"),
+        widget=forms.Select(choices=[]),
+    )
+    name = forms.CharField(
+        label=ugettext_lazy("Public Name"),
+        help_text=ugettext_lazy(
+            "Users will see this name when logging in."
+        ),
+        max_length=128,
+    )
+    slug = forms.CharField(
+        label=ugettext_lazy("Slug for SP Endpoints"),
+        help_text=ugettext_lazy(
+            "This will be the unique slug for this IdP's SAML2 urls. "
+        ),
+        max_length=256,
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.helper = FormHelper()
+        self.helper.label_class = 'col-sm-3 col-md-2'
+        self.helper.field_class = 'col-sm-9 col-md-8 col-lg-6'
+        self.helper.layout = crispy.Layout(
+            crispy.Fieldset(
+                _('Primary Configuration'),
+                crispy.Field(
+                    'owner',
+                    css_class="input-xxlarge",
+                    placeholder="Search for Billing Account"
+                ),
+                'name',
+                'slug',
+            ),
+            hqcrispy.FormActions(
+                twbscrispy.StrictButton(
+                    ugettext_lazy("Create Identity Provider"),
+                    type="submit",
+                    css_class="btn btn-primary",
+                )
+            )
+        )
+
+    def clean_slug(self):
+        slug = self.cleaned_data['slug']
+        _validate_or_raise_slugify_error(slug)
+        if IdentityProvider.objects.filter(slug=slug).exists():
+            raise forms.ValidationError(
+                _("A Identity Provider already exists with this slug.")
+            )
+        return slug
+
+    @transaction.atomic
+    def create_identity_provider(self, admin_user):
+        owner = BillingAccount.objects.get(id=self.cleaned_data['owner'])
```

Comment:
i'm a little confused? `is_valid` should be called in the view...

Reply:
Sorry, I basically mean defensive programming. With a method called `create_identity_provider()`, it's reasonable to assume all you need to do is create a form, supply data to the fields, and then call the method.  The implicit knowledge is that the user is expected to know that someone needs to have called `is_valid()` on that form first, because attempting to access these `self.cleaned_data` fields will raise an exception, otherwise. To prevent this, you could do something like the following:

```
def create_identity_provider(self, admin_user):
  if not self.is_valid():
    return None # or raise exception
  owner = BillingAccount.objects.get(...)
```

Now the class can't be misused. The downside is we're making an extra `is_valid()` call, but the documentation suggests this doesn't do anything if its already been run.

Reply:
i get what you are saying. I'm just wondering if it's really needed considering the
```
if form.is_valid():
   # do something to process it
   return redirect("foo")
```

Is a pretty standard pattern when dealing with django forms and I don't think there would be any expectation to call `create_identity_provider` on a form without following that pattern?